pkg/errors/user_errors.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errors
+
+import (
+	"errors"
+)
+
+type UserMessenger interface {
+	error
+	UserMessage() string
+}
+
+// UserError is an error type that be checked to be present in the
+// error chain. And if present, can be shown to the end users
+type UserError struct {
+	error
+	msg string
+}
+
+func (ue *UserError) UserMessage() string {
+	return ue.msg
+}
+
+func (ue *UserError) Unwrap() error { return ue.error }
+
+// UserErrorWithMessage initializes the UserError with given error and a message.
+// Needs to accept error if we want to initialize the UserError in between the
+// error chain
+func UserErrorWithMessage(err error, msg string) UserMessenger {
+	if err == nil {
+		err = errors.New(msg)
+	}
+	return &UserError{err, msg}
+}
+
+// UserMessagesInError gets us messages of all the `UserError`s that are wrapped
+// in the passed error's chain
+func UserMessagesInError(err error) []string {
+	userMessages := []string{}
+	if err == nil {
+		return nil
+	}
+	for err != nil {
+		if e, ok := err.(*UserError); ok {
+			userMessages = append(userMessages, e.UserMessage())
+		}
+		err = errors.Unwrap(err)
```

Comment:
Is this reliable? What if there is an intermediate error in the chain that didn't implement `Unwrap()`, causing ` err == nil`? Any subsequent `UserError`s will be lost.

Reply:
Interesting, how come there can be an error in the chain, which doesn't implement `Unwrap`, but at the same time wraps another error? I can imagine that only as a consequence of a potential bug (a struct that implements `error`, that allows wrapping an error, but someone forgot to implement `Unwrap` for that struct).

Reply:
mm, good points. Let me try to writer some test around it to see the behaviour.

Reply:
But if the downstream custom error does not implement it, it means they (be it intentional or not) do not provide any means of getting the underlying error. Is there anything we can do about it?

Reply:
See my suggested code change for ways to get rid of the `for` loop.